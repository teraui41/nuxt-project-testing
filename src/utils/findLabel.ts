import { Permission, Permissions } from "../types/Permissions";

const PermissionReducer = (parent, level) => (pre: object, permission: Permission) => {
  const { children, code, label } = permission;

  const morePermissionItems = children.length > 0
    ? flatPermissionList(children, code, level + 1)
    : {};

  return { ...pre, [code]: { code, label, parent, level}, ...morePermissionItems};
};

export function flatPermissionList(permissions: Permissions, parent: string = "", level = 0): object {
  const FlattenPermission = permissions.reduce(PermissionReducer(parent, level), {});
  return FlattenPermission;
}

export function findLabel(code: string, permissions: Permissions): string {

  // 將權限表攤平
  const flatList = flatPermissionList(permissions);

  const result: any = flatList[code];

  return result !== null
    ? result.label
    : "";
}
