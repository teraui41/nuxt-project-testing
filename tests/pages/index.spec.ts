import { mount } from "@vue/test-utils";
import {mockPermissions} from "../../dummies/mockPermissions";
import Index from "../../src/pages/index.vue";
import { findLabel } from "../../src/utils/findLabel";

const testList = [
  {
    value: "code A",
    expectedValue: "label A",
  },
  {
    value: "code B",
    expectedValue: "label B",
  },
  {
    value: "code C",
    expectedValue: "label C",
  },
  {
    value: "code B-1",
    expectedValue: "label B-1",
  },
  {
    value: "code B-1-1",
    expectedValue: "label B-1-1",
  },
  {
    value: "code B-2",
    expectedValue: "label B-2",
  },
  {
    value: "code C-1",
    expectedValue: "label C-1",
  },
  {
    value: "code E",
    expectedValue: "",
  },
];

describe("index spec", () => {

  it("should render proper", () => {
    const wrapper = mount(Index);
    expect(wrapper.element).toMatchSnapshot();
  });

  testList.forEach( (testItem) => {
    it(`validate code ${testItem.value}`, () => {
      const result = findLabel(testItem.value, mockPermissions);
      expect(result).toBe(testItem.expectedValue);
    });
  });

});
