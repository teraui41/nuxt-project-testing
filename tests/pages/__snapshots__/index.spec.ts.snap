// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index spec should render proper 1`] = `
<div
  class="bg"
>
  <div
    class="container panel"
  >
    <h1>
      Edmond Wu
    </h1>
     
    <div
      class="row"
    >
      <div
        class="col-12 col-lg-6"
      >
        <div
          class="card"
        >
          <div
            class="card-body"
          >
            <h5
              class="card-title"
            >
              Find Label Function Block
            </h5>
             
            <div
              class="form-group col-md-6"
            >
              <label>
                Search Code Input
              </label>
               
              <input
                class="form-control"
                placeholder="please input the permission code"
                type="text"
              />
            </div>
             
            <div
              class="form-group col-md-6"
            >
              <label>
                Result Code
              </label>
               
              <input
                class="form-control"
                placeholder="search result"
                readonly=""
                type="text"
              />
            </div>
          </div>
        </div>
      </div>
       
      <div
        class="col-12 col-lg-6"
      >
        <div
          class="card"
        >
          <div
            class="card-body"
          >
            <h5>
              Tree View
            </h5>
             
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
